### PR TITLE
Development

### PR DESCRIPTION
**Pull request by htuerker**
tic-tac-toe project's pull request is below there.
[Tic-Tac-Toe](https://github.com/btuerker/tic-tac-toe/pull/2)

### CODE REVIEW

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    @variable = variable
+    @other_variable = other_variable
+  end
+end
+
+RSpec.describe Enumerable do
+  describe "#my_each" do
+    it "should access all element in given array" do
+      counter = 0
+      array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```
Comment (heyapricot) [inline spec/enumerable_spec.rb]: [Use let or let! when defining test subjects ](http://www.betterspecs.org/#let)

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,160 @@
+require('./lib/enumerable.rb')
+include Enumerable
+
+class MockObject
+  attr_reader :variable, :other_variable
+  def initialize variable = 'hello', other_variable = 'world'
+    @variable = variable
+    @other_variable = other_variable
+  end
+end
```
Comment (heyapricot) [inline spec/enumerable_spec.rb]: What's the idea behind this?

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        expect(e.variable).to eql('hello')
+        expect(e.other_variable).to eql('world')
+      end
+    end
+
+  end
+  describe "#my_each_with_index" do
+    it "should return an element by index" do
+      array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+      expect(array.my_each_with_index(0)).to eql(1)
```
Comment (heyapricot) [inline spec/enumerable_spec.rb]: This is not how each_with_index works. Take a look [here](https://ruby-doc.org/core-2.6/Enumerable.html#method-i-each_with_index)

File: spec/enumerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        backward_counter -= 1
+      end
+    end
+  end
+
+  describe "#my_select" do
+    it "should return a new list that contains elements that matches with the given block condition" do
+      numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+      even_numbers = numbers.my_select { |e| e.even? }
+      expect(even_numbers).to be_an_instance_of(Array)
```
Comment (heyapricot) [inline spec/enumerable_spec.rb:32]: Nice touch 👍🏼

File: spec/enumerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      expect(even_numbers).to eql([2, 4, 6, 8, 10])
+    end
+  end
+
+  describe "#my_all?" do
+    it "should return a boolean" do
+      expect([1, 2, 3].my_all? { }).to be(true).or be(false)
```
Comment (heyapricot) [inline spec/enumerable_spec.rb:39]: Could have checked instance of TrueClass or FalseClass. Also nice haha

File: spec/enumerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  describe "#my_count" do
+    it "should return 0 when array is empty" do
+      empty_array = []
+      expect(empty_array.my_count).to eql(0)
+    end
+
+    it "should return size of array" do
+      array = [1,2,3,4,5]
+      expect(array.my_count).to eql(array.length)
+      expect(array.my_count).to eql(5)
```
Comment (heyapricot) [inline spec/enumerable_spec.rb]: I think you are fine checking just against array.length

File: spec/enumerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    end
+  end
+
+  describe "#my_map" do
+    it "should return a new array processed by the given block" do
+      numbers = [1,2,3,4,5]
+      numbers_powed = numbers.my_map { |e| e = e * e }
```
Comment (heyapricot) [inline spec/enumerable_spec.rb]: You don't need to assign the `e` variable. Map makes an array with the returned value (And Ruby automatically returns the last executed value). You could also use exponent operator **. `numbers.my_map { |e| e ** 2 }`

File: spec/enumerable_spec.rb [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      expect(numbers_powed).to be_instance_of(Array)
+      expect(numbers_powed).to eql([1,4,9,16,25])
+    end
+  end
+
+  describe "#my_inject" do
+    it "should make calculation memoizated value" do
+      numbers = [1,2,3,4,5]
+      sum = numbers.inject { |memo, e| memo + e }
+      expect(sum).to eql(15)
```
Comment (heyapricot) [inline spec/enumerable_spec.rb:107]: Check against the original `inject` instead of hardcoding the value

File: spec/enumerable_spec.rb [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+      expect(array.my_count).to eql(5)
+    end
+  end
+
+  describe "#my_map" do
+    it "should return a new array processed by the given block" do
+      numbers = [1,2,3,4,5]
+      numbers_powed = numbers.my_map { |e| e = e * e }
+      expect(numbers_powed).to be_instance_of(Array)
+      expect(numbers_powed).to eql([1,4,9,16,25])
```
Comment (heyapricot) [inline spec/enumerable_spec.rb:99]: You can check against the original `map`

**Review by heyapricot — CHANGES_REQUESTED** [review]
# Just one thing to change

👋🏼 Hello! Overall the project looks good. My only concern is that the test for each_index is not right. Please fix it and after that send a new code review request. Thanks! :)

File: spec/enumerable_spec.rb (hunk 1 above)
Comment (htuerker) [reply spec/enumerable_spec.rb]: We used it but as I mentioned in tic-tac-toe, I think that this idea doesn't fit at all for this project. If we use some service layers as a object something like FooCalculatorService, we can use this declarations but let me say that we need to specific arrays for some scenarios for example positive_numbers or negative_numbers

File: spec/enumerable_spec.rb (hunk 2 above)
Comment (htuerker) [reply spec/enumerable_spec.rb]: :) Just mocking and try to test our module can work with objects. But I know that everything is object in ruby :)

File: spec/enumerable_spec.rb (hunk 4 above)
Comment (htuerker) [reply spec/enumerable_spec.rb:32]: Thank you :)

File: spec/enumerable_spec.rb (hunk 6 above)
Comment (htuerker) [reply spec/enumerable_spec.rb]: Changed it

File: spec/enumerable_spec.rb (hunk 3 above)
Comment (htuerker) [reply spec/enumerable_spec.rb]: We didn't get the concept behind each_with_index. It was changed and ready to review by you!

File: spec/enumerable_spec.rb (hunk 5 above)
Comment (htuerker) [reply spec/enumerable_spec.rb:39]: To be or not to be :) That's the question

File: spec/enumerable_spec.rb (hunk 7 above)
Comment (htuerker) [reply spec/enumerable_spec.rb]: Nice trick, actually we know it but we forgot to use every time

File: spec/enumerable_spec.rb (hunk 5 above)
Comment (heyapricot) [reply spec/enumerable_spec.rb:39]: Yes, I had to do some research about the boolean type on Ruby. Just wanted to share the results of my research but again, I think you solved it nicely :)

**Review by heyapricot — APPROVED** [review]
# Nice work 👍🏼
Good job with the testing. No need from my side for further review. You can move on 😬

**Pull request merged**